GSB: Lazy requirement signature

xymus opened the pull request:
Make getRequirementSignature lazy for the signature to be computed only on its first use. The requirement signature is still explicitly computed on the primary files to preserve diagnostics.

xymus committed:
Explicitly request to createGenericParamsIfMissing as it will be lazy

Computing the requirement signature created the generic params as
a side effect. Making getRequirementSignature lazy means that users
of the generic params must make sure they are created before use.

xymus committed:
Intro a request to compute the requirement signature of a ProtocolDecl

Replaces the explicit call to computeRequirementSignature from
validateDecl with a lazy getRequirementSignature. A side effect is that
the generic params of a ProtocolDecl are no longer computed from
validateDecl and must be computed lazily too.

xymus committed:
Remove explicit calls to computeRequirementSignature

We can rely on lazy calls to getRequirementSignature instead.

slavapestov commented on lib/Sema/TypeCheckProtocolInference.cpp:
```diff
@@ -2000,6 +2000,9 @@ void ConformanceChecker::resolveTypeWitnesses() {
   };
 
+  // Force computing the requirement signature first.
+  (void) Proto->getRequirementSignature();
```
This should not be needed.

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1384,6 +1384,13 @@ IsDynamicRequest::evaluate(Evaluator &evaluator, ValueDecl *decl) const {
 
+llvm::Expected<bool>
+RequirementSignatureRequest::evaluate(Evaluator &evaluator, ProtocolDecl *proto) const {
+  const_cast<ProtocolDecl*>(proto)->computeRequirementSignature();
```
You should fold computeRequirementSignature() here.

slavapestov commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -243,6 +243,32 @@ class IsDynamicRequest :
+class RequirementSignatureRequest :
+    public SimpleRequest<RequirementSignatureRequest,
+                         CacheKind::Uncached,
+                         bool, // FIXME valid result kind
```
You should figure out how to store an ArrayRef here

slavapestov commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -243,6 +243,32 @@ class IsDynamicRequest :
+/// Compute the requirements that describe a protocol.
+class RequirementSignatureRequest :
+    public SimpleRequest<RequirementSignatureRequest,
+                         CacheKind::Uncached,
```
Should be ::Cached or ::SeparatelyCached

slavapestov reviewed (commented):
This looks like a good start. Can you rip out all the existing explicit calls to computeRequirementSignature() too?

slavapestov commented:
... also, any callers of `ProtocolDecl::isRequirementSignatureComputed()` should be removed or refactored. There's a tricky one in GenericSignatureBuilder; there, the right check you want is not "have I computed a requirement signature yet", but "am I currently not already computing a requirement signature". @DougGregor can help you implement that in terms of the request evaluator itself

xymus committed:
Don’t use isRequirementSignatureComputed because it is lazy

xymus committed:
Intro Evaluator::hasActiveRequest to prevent request cycles

xymus committed:
Avoid cycles when computing a requirement signature

Replace most remaining uses of isRequirementSignatureComputed by
isRequirementSignatureComputing which uses Evaluator::hasActiveRequest
to detect if the requirements are currently being computed.

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

slavapestov commented on include/swift/AST/Decl.h:
```diff
@@ -4302,22 +4303,22 @@ class ProtocolDecl final : public NominalTypeDecl {
+  ArrayRef<Requirement> getRequirementSignature() const;
+
+  /// Is the requirement signature currently being computed?
+  bool isRequirementSignatureComputing() const;
```
Nitpick: isComputingRequirementSignature() reads better perhaps

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -4495,9 +4527,7 @@ static void finalizeType(TypeChecker &TC, NominalTypeDecl *nominal) {
-    if (!PD->isRequirementSignatureComputed()) {
-      TC.validateDecl(PD);
-    }
+    TC.validateDecl(PD);
```
This might not be needed at all

slavapestov commented on lib/Sema/TypeCheckProtocolInference.cpp:
```diff
@@ -1095,9 +1095,6 @@ AssociatedTypeInference::getSubstOptionsWithCurrentTypeWitnesses() {
-  // If we don't have a requirement signature for this protocol, bail out.
-  // FIXME: We should never get to this point. Or we should always fail.
-  if (!proto->isRequirementSignatureComputed()) return false;
 
```
Remove blank line

slavapestov commented on lib/Serialization/Serialization.cpp:
```diff
@@ -3284,6 +3284,7 @@ void Serializer::writeDecl(const Decl *D) {
                                inherited);
 
+    const_cast<ProtocolDecl*>(proto)->createGenericParamsIfMissing();
     writeGenericParams(proto->getGenericParams());
```
In a separate PR can you try moving the createGenericParamsIfMissing() call into getGenericParams() itself?

slavapestov reviewed (approved):
Looks great! You can address the nitpicks in a followup.

slavapestov commented:
I think the next step is to try to remove TypeChecker::DelayedRequirementSignatures. It should not be needed any more.

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

DougGregor commented on lib/AST/ASTDumper.cpp:
```diff
@@ -708,13 +708,9 @@ namespace {
-        OS << "<null>";
-      }
+      OS << GenericSignature::get({PD->getProtocolSelfType()} ,
+                                   PD->getRequirementSignature())
```
This is the one place I can think of where we still want to check `isRequirementSignatureComputed()`, because we want the AST dumper to dump only what's currently been realized in the AST... and avoid performing extra computation (in this case, a bunch of type checking) that makes it harder to debug.

xymus replied:
Alright, I brought back the original version.

DougGregor reviewed (approved):
This looks great! Just the one comment about the AST dumper.

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility
